clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2Credentials.java

```diff
@@ -679,4 +573,130 @@ public void patch(
+      // a long time); this flag was added to skip the checks and assume all kinds were readable.
+      // Now that permissions are checked on-the-fly, this flag is probably not necessary, but for
+      // now we'll continue to support the prior behavior, which is to short-circuit and assume all
+      // kinds are readable before checking.
+      return !checkPermissionsOnStartup;
+    }
+
+    private boolean canReadKind(KubernetesKind kind) {
+      if (skipPermissionChecks()) {
+        return true;
+      }
+      log.info("Checking if {} is readable in account '{}'...", kind, accountName);
+      boolean allowed;
+      if (kindRegistry.getRegisteredKind(kind).isNamespaced()) {
+        allowed =
+            jobExecutor.authCanINamespaced(
+                KubernetesV2Credentials.this, getCheckNamespace(), kind.getName(), "list");
+      } else {
+        allowed = jobExecutor.authCanI(KubernetesV2Credentials.this, kind.getName(), "list");
+      }
+
+      if (!allowed) {
+        log.info(
+            "Kind {} will not be cached in account '{}' because it cannot be listed.",
+            kind,
+            accountName);
+      }
+
+      return allowed;
+    }
+
+    private boolean checkMetricsReadable() {
+      if (skipPermissionChecks()) {
+        return true;
+      }
+      try {
+        log.info("Checking if pod metrics are readable for account {}...", accountName);
+        topPod(getCheckNamespace(), null);
+        return true;
+      } catch (Exception e) {
+        log.warn(
+            "Could not read pod metrics in account '{}' for reason: {}",
+            accountName,
+            e.getMessage());
+        log.debug("Reading logs for account '{}' failed with exception: ", accountName, e);
+        return false;
+      }
+    }
+
+    /**
+     * Returns the status of a given kubernetes kind with respect to the current account. Checks of
+     * whether a kind is readable are cached for the lifetime of the process (and are only performed
+     * when a kind is otherwise considered valid for the account).
+     */
+    @Nonnull
+    KubernetesKindStatus getKindStatus(@Nonnull KubernetesKind kind) {
+      if (kind.equals(KubernetesKind.NONE)) {
+        return KubernetesKindStatus.KIND_NONE;
+      }
+
+      if (!kinds.isEmpty()) {
+        return kinds.contains(kind)
+            ? KubernetesKindStatus.VALID
+            : KubernetesKindStatus.MISSING_FROM_ALLOWED_KINDS;
+      }
+
+      if (omitKinds.contains(kind)) {
+        return KubernetesKindStatus.EXPLICITLY_OMITTED_BY_CONFIGURATION;
+      }
+
+      return readableKinds.computeIfAbsent(kind, this::canReadKind)
+          ? KubernetesKindStatus.VALID
+          : KubernetesKindStatus.READ_ERROR;
+    }
+
+    /**
+     * Returns whether metrics are readable for the current kubernetes account. This check is cached
+     * for the lifetime of the process, and subsequent calls return the cached value.
+     */
+    boolean isMetricsReadable() {
```

Comment:
Putting this metrics stuff in a class called `KindValidator` is sort of confusing (it seems to be entirely unrelated to validating kinds?) but not a huge deal, I guess...

Reply:
Would it be more clear if this were called something like `PermissionValidator`?  While `metrics` is not really a kind, the thing we're doing is roughly the same---checking to see if the current service account has permission to read something in the cluster.

Reply:
Yeah, I think it would.